influxql/call_iterator.go

```diff
@@ -362,18 +360,30 @@ func booleanLastReduce(prev, curr *BooleanPoint) (int64, bool, []interface{}) {
 func NewDistinctIterator(input Iterator, opt IteratorOptions) (Iterator, error) {
 	switch input := input.(type) {
 	case FloatIterator:
-		return &floatReduceSliceIterator{input: newBufFloatIterator(input), opt: opt, fn: floatDistinctReduceSlice}, nil
+		createFn := func() (FloatPointSliceAggregator, FloatPointSliceEmitter) {
+			fn := NewFloatSliceFuncReducer(FloatDistinctReduceSlice)
+			return fn, fn
+		}
+		return &floatReduceSliceIterator{input: newBufFloatIterator(input), opt: opt, create: createFn}, nil
 	case IntegerIterator:
-		return &integerReduceSliceIterator{input: newBufIntegerIterator(input), opt: opt, fn: integerDistinctReduceSlice}, nil
+		createFn := func() (IntegerPointSliceAggregator, IntegerPointSliceEmitter) {
+			fn := NewIntegerSliceFuncReducer(IntegerDistinctReduceSlice)
+			return fn, fn
+		}
+		return &integerReduceSliceIterator{input: newBufIntegerIterator(input), opt: opt, create: createFn}, nil
 	case StringIterator:
-		return &stringReduceSliceIterator{input: newBufStringIterator(input), opt: opt, fn: stringDistinctReduceSlice}, nil
+		createFn := func() (StringPointSliceAggregator, StringPointSliceEmitter) {
+			fn := NewStringSliceFuncReducer(StringDistinctReduceSlice)
+			return fn, fn
+		}
+		return &stringReduceSliceIterator{input: newBufStringIterator(input), opt: opt, create: createFn}, nil
 	default:
 		return nil, fmt.Errorf("unsupported distinct iterator type: %T", input)
 	}
 }
 
-// floatDistinctReduceSlice returns the distinct value within a window.
-func floatDistinctReduceSlice(a []FloatPoint, opt *reduceOptions) []FloatPoint {
+// FloatDistinctReduceSlice returns the distinct value within a window.
+func FloatDistinctReduceSlice(a []FloatPoint, opt *ReduceOptions) []FloatPoint {
```

Comment:
When you say ZeroTime do you mean the int64 == 0 or the constant ZeroTime?
If its the later then its the same amount of work to set the time to the startTime.